Implementing cross-platform support by supporting Linux-friendly HTTP clients

mergesort opened the pull request:
This PR adds support for Linux, as requested in #89. It does so by:
- Creating a new networking layer using Apple's [AsyncHTTPClient](https://github.com/swift-server/async-http-client) library. This is considered to be the gold-standard for Swift on Linux, and is also a solid networking layer for iOS/macOS apps, and fixes the many bugs that URLSession has on Linux platforms.
- To ensure compatibility on Apple platforms, I've added a URLSessionHTTPClientAdapter which makes network requests through URLSession.
- When creating a new service, there is a configurable parameter for users to specify their preferred HTTPClient. This allows maximum flexibility for end-users, in case they have a preference for networking stack.

> [!NOTE]
>These changes are not applied to any of the code related to AIProxy. I've left everything alone there as best as I can, and simply do not compile any AIProxy code on Linux, since AIProxy does not make much sense to use in a server environment where API requests are already protected.

Please let me know what you think, happy to make any changes you request. Would love to get this merged in so I can start using SwiftOpenAI in my server-side app!

gsabran commented on Package.swift:
```diff
@@ -16,11 +16,17 @@ let package = Package(
       name: "SwiftOpenAI",
       targets: ["SwiftOpenAI"]),
   ],
+  dependencies: [
+    .package(url: "https://github.com/swift-server/async-http-client.git", from: "1.25.2"),
+  ],
```
I'm not sure this is possible, but can we scope this to only linux for now? ie wrap in `#if os(Linux)` ?

mergesort replied:
I've made this change by using `condition: .when(platforms: [.linux])`, since `#if os(Linux)` doesn't work in Package.swift.

gsabran commented on Sources/OpenAI/AIProxy/AIProxyService.swift:
```diff
@@ -44,8 +44,10 @@ struct AIProxyService: OpenAIService {
+    self.httpClient = NoOpClient()
   }
 
+  let httpClient: HTTPClient
   let session: URLSession
```
I'm not super familiar with the code, but it seems weird to see both  HTTPClient and URLSession properties here.

gsabran replied:
why not remove the `session` property, the `NoOpClient` type, and do
```swift
self.httpClient = URLSessionHTTPClientAdapter(urlSession: URLSession(
      configuration: .default,
      delegate: aiproxySecureDelegate,
      delegateQueue: nil))
```

mergesort replied:
This was a good idea, removed the `NoOpClient` in favor of your solution!

gsabran commented on Sources/OpenAI/Private/Networking/AsyncHTTPClientAdapter.swift:
```diff
@@ -0,0 +1,131 @@
+import NIOHTTP1
+
+/// Adapter that implements HTTPClient protocol using AsyncHTTPClient
+public class AsyncHTTPClientAdapter: HTTPClient {
```
Maybe to reduce the scope of the change in this PR, what do you think of:
- scope this code to only Linux.
- replaces usages of `AsyncHTTPClientAdapter.createDefault()` by `HTTPClient.createDefault()`
- add
```swift
extension HTTPClient {
  static func createDefault() -> HttpClient {
    #if os(Linux)
    AsyncHTTPClientAdapter.createDefault()
    #else
    URLSessionHTTPClientAdapter()
    #endif
  }
```

gsabran commented on Sources/OpenAI/Public/Service/OpenAIServiceFactory.swift:
```diff
@@ -14,23 +14,24 @@ public class OpenAIServiceFactory {
   ///   - organizationID: The optional organization ID for multi-tenancy (default is `nil`).
-  ///   - configuration: The URL session configuration to be used for network calls (default is `.default`).
   ///   - decoder: The JSON decoder to be used for parsing API responses (default is `JSONDecoder.init()`).
+  ///   - httpClient: The HTTPClient to be used for network calls. Defaults to `AsyncHTTPClientAdapter.createDefault()`
```
@jamesrochabrun how do you feel about changing existing APIs here?

Would you prefer to keep the existing ones available outside of Linux, and have them map to the new ones, at the cost of having two functions that mostly do the same thing, or are you fine with the breaking change?

jamesrochabrun replied:
I think its better to have it as another initializer, I would like to avoid breaking changes if possible.

gsabran commented on Sources/OpenAI/Public/Service/OpenAIService.swift:
```diff
@@ -1005,44 +1007,47 @@ extension OpenAIService {
+
+      // Convert URLRequest to HTTPRequest
+      let httpRequest = HTTPRequest(
+        url: request.url!,
```
the `!` always looke scary

gsabran commented on Sources/OpenAI/Public/Service/OpenAIService.swift:
```diff
@@ -1005,44 +1007,47 @@ extension OpenAIService {
-    guard let httpResponse = response as? HTTPURLResponse else {
-      throw APIError.requestFailed(description: "invalid response unable to get a valid HTTPURLResponse")
-    }
-    printHTTPURLResponse(httpResponse)
```
note: this is removed

gsabran commented on Sources/OpenAI/Public/Service/OpenAIService.swift:
```diff
@@ -1157,258 +1173,253 @@ extension OpenAIService {
-        // If decoding fails, proceed with a general error message
-        errorMessage = "status code \(httpResponse.statusCode)"
+      // Convert URLRequest to HTTPRequest
+      let httpRequest = HTTPRequest(
+        url: request.url!,
+        method: HTTPMethod(rawValue: request.httpMethod ?? "GET") ?? .get,
+        headers: request.allHTTPHeaderFields ?? [:],
+        body: request.httpBody
+      )
```
nit: this is repeated several times. Maybe add an initializer in `HTTPRequest` from `URLRequest` ?

gsabran reviewed (commented):
I left a few comments, mostly with the goal to not change anything on non Linux platforms other than adding a layer of indirection that is fine.

If this merges, you might want to follow up with adding some CI on Linux (maybe just build and run tests) to prevent future breakages as I don't think a lot of people would think about testing on this platform.

mergesort commented:
These are all great suggestions, I'll address them in the next couple of days.

I should have noted this but I had left in the AsyncHTTPClient code on iOS/macOS so changes could be tested by people on Apple platforms, but happy to contain everything AsyncHTTPClient-related to Linux only.

jamesrochabrun commented:
Thanks @mergesort for your willingness to address @gsabran feedback!

mergesort commented:
Sorry for the time it took, but I believe I've addressed all of the concerns. Please let me know if I've missed anything — excited to have this functionality on the server!

jamesrochabrun commented:
@mergesort this is awesome, thanks for addressing the feedback, I will defer to @gsabran to do the final review, but can I please ask you to add a section in the Read me in how to use this new functionality? it will be nice to show it for others. Thanks!

mergesort commented:
@AlekseyPleshkov I'm so sorry about that, I had been testing against my app's server, which I plan to deploy this on. I thought everything had worked on Linux, but only now did I realize that in my haste to push a new feature last week I had commented out an internal import of the package containing SwiftOpenAI, so I was seeing successful builds when they would have failed.

I've created a docker image locally for testing, and am once again trying to resolve the issues you're seeing. You may see a few pushes from me, especially as I work through adding Linux testing on CI (CI is not a strength of mine), so please excuse me if that doesn't work yet and you get a few notifications about me pushing changes.

@jamesrochabrun Yes, will absolutely add some notes to the readme before considering this done. 🙂

jamesrochabrun commented:
@AlekseyPleshkov do you mind testing again with latest updates on this branch? Planning to merge this soon thanks
